File: src/main/kotlin/com/airsoft/tactic/admin/service/AdminAuthService.kt
package com.airsoft.tactic.admin.service

import com.airsoft.tactic.admin.dto.AdminInfo
import com.airsoft.tactic.admin.dto.AdminLoginRequest
import com.airsoft.tactic.admin.dto.AdminLoginResponse
import com.airsoft.tactic.admin.dto.CreateAdminRequest
import com.airsoft.tactic.entity.AdminAccount
import com.airsoft.tactic.exception.AppException
import com.airsoft.tactic.repository.AdminAccountRepository
import com.airsoft.tactic.util.JwtUtil
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.stereotype.Service
import java.time.Instant

@Service
class AdminAuthService(
    private val adminAccountRepository: AdminAccountRepository,
    private val jwtUtil: JwtUtil,
    private val passwordEncoder: BCryptPasswordEncoder
) {
    fun login(req: AdminLoginRequest): AdminLoginResponse {
        val account = adminAccountRepository.findByUsername(req.username)
            ?: throw AppException.unauthorized("TOKEN_INVALID", "Invalid username or password")

        if (!account.isActive)
            throw AppException.unauthorized("TOKEN_INVALID", "Account is disabled")

        if (!passwordEncoder.matches(req.password, account.passwordHash))
            throw AppException.unauthorized("TOKEN_INVALID", "Invalid username or password")

        account.lastLoginAt = Instant.now()
        adminAccountRepository.save(account)

        val token = jwtUtil.generateAccessToken(account.id!!, account.username, "admin")
        return AdminLoginResponse(
            accessToken = token,
            expiresIn = jwtUtil.accessTokenValiditySeconds,
            admin = AdminInfo(
                id = account.id!!.toString(),
                username = account.username,
                displayName = account.displayName
            )
        )
    }

    fun createAdmin(req: CreateAdminRequest): AdminInfo {
        if (adminAccountRepository.findByUsername(req.username) != null)
            throw AppException.conflict("ALREADY_EXISTS", "Username '${req.username}' already taken")

        val account = adminAccountRepository.save(
            AdminAccount(
                username = req.username,
                passwordHash = passwordEncoder.encode(req.password),
                displayName = req.displayName.ifBlank { "Admin" }
            )
        )
        return AdminInfo(
            id = account.id!!.toString(),
            username = account.username,
            displayName = account.displayName
        )
    }
}

File: src/admin/admin-auth.service.ts
import {
  ConflictException,
  Injectable,
  Logger,
  OnApplicationBootstrap,
  UnauthorizedException,
} from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import * as bcrypt from 'bcrypt';
import { AdminAccount } from '../database/entities/admin-account.entity';
import { JwtUtil } from '../common/utils/jwt.util';
import { AdminLoginDto } from './dto/admin-login.dto';
import { CreateAdminDto } from './dto/create-admin.dto';
import { AdminInfoDto, AdminLoginResponseDto } from './dto/admin-login-response.dto';

const BCRYPT_ROUNDS = 12;

@Injectable()
export class AdminAuthService implements OnApplicationBootstrap {
  private readonly logger = new Logger(AdminAuthService.name);

  constructor(
    @InjectRepository(AdminAccount)
    private readonly adminAccountRepo: Repository<AdminAccount>,
    private readonly jwtUtil: JwtUtil,
  ) {}

  async onApplicationBootstrap(): Promise<void> {
    // Auto-bootstrap on startup is intentionally skipped here.
    // Call POST /v1/admin/auth/bootstrap explicitly instead.
  }

  async login(dto: AdminLoginDto): Promise<AdminLoginResponseDto> {
    const account = await this.adminAccountRepo.findOne({
      where: { username: dto.username },
    });

    if (!account) {
      throw new UnauthorizedException('Invalid username or password');
    }

    if (!account.isActive) {
      throw new UnauthorizedException('Account is disabled');
    }

    const passwordMatch = await bcrypt.compare(dto.password, account.passwordHash);
    if (!passwordMatch) {
      throw new UnauthorizedException('Invalid username or password');
    }

    account.lastLoginAt = new Date();
    await this.adminAccountRepo.save(account);

    const accessToken = this.jwtUtil.sign({
      sub: account.id,
      email: account.username,
      role: 'admin',
    });

    const adminInfo: AdminInfoDto = {
      id: account.id,
      username: account.username,
      displayName: account.displayName,
    };

    return {
      accessToken,
      expiresIn: this.jwtUtil.accessTokenExpiresIn,
      admin: adminInfo,
    };
  }

  async bootstrap(): Promise<AdminInfoDto> {
    const defaultPassword = 'Admin123@';
    const existing = await this.adminAccountRepo.findOne({
      where: { username: 'admin' },
    });

    if (existing) {
      // Always reset password to default
      existing.passwordHash = await bcrypt.hash(defaultPassword, BCRYPT_ROUNDS);
      await this.adminAccountRepo.save(existing);
      this.logger.log('Bootstrap: reset admin password to default');
      return { id: existing.id, username: existing.username, displayName: existing.displayName };
    }

    return this.createAdmin({
      username: 'admin',
      password: defaultPassword,
      displayName: 'System Admin',
    });
  }

  async createAdmin(dto: CreateAdminDto): Promise<AdminInfoDto> {
    const existing = await this.adminAccountRepo.findOne({
      where: { username: dto.username },
    });

    if (existing) {
      throw new ConflictException(`Username '${dto.username}' already taken`);
    }

    const passwordHash = await bcrypt.hash(dto.password, BCRYPT_ROUNDS);
    const displayName = dto.displayName.trim() || 'Admin';

    const account = this.adminAccountRepo.create({
      username: dto.username,
      passwordHash,
      displayName,
    });
    const saved = await this.adminAccountRepo.save(account);

    return {
      id: saved.id,
      username: saved.username,
      displayName: saved.displayName,
    };
  }
}
